ENH : issue/16번 RabbitMQ DLQ 재처리 로직 추가

gamsayeon opened the pull request.

gamsayeon committed:
BUG: ModelMapping 매핑 전략이 LOOSE에서 STRICT으로 변경(USERDTO TO USER에서 매핑이 안되는 문제 해결)

gamsayeon committed:
ENH: 엔드 포인트 restful하게 수정 및 BidPriceValidServiceImpl 중복코드 제거와 리펙토링

ccommit-dev commented on src/main/java/com/ccommit/auction_server/service/serviceImpl/BidPriceValidServiceImpl.java:
```diff
@@ -25,43 +26,32 @@ public class BidPriceValidServiceImpl implements BidPriceValidService {
         Product product = productRepository.findByProductId(productId);
         Optional<Category> category = categoryRepository.findByCategoryId(product.getCategoryId());
-        int minBidPriceUnit = category.get().getBidMinPrice();
+        int categoryMinimumBidPrice = category.get().getBidMinPrice();
```
@gamsayeon  카테고리별 최저 입찰가가 필요한 이유를 주석으로 추가해주세요

gamsayeon replied:
무분별한 입찰을 방지 하기 위한 최저 입찰가로 최저 입찰가가 없으면, 1원단위의 무분별한 입찰이 일어날 수 있습니다. 이를 방지 하기 위함입니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/service/serviceImpl/BidPriceValidServiceImpl.java:
```diff
@@ -25,43 +26,32 @@ public class BidPriceValidServiceImpl implements BidPriceValidService {
+        int productHighestPrice = product.getHighestPrice();
+        if (currentHighestBid == null) {
+            int startingPrice = product.getStartPrice();
+            if (newBidPrice > productHighestPrice || newBidPrice < startingPrice ||
```
startingPrice 변수명이 시작 입찰가라는 의미를 모두 담게 변수명을 수정 부탁드립니다.
EX) startingBidPrice

gamsayeon replied:
넵 변수명을 다시 고려하여 수정하도록 하겠습니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/service/serviceImpl/BidPriceValidServiceImpl.java:
```diff
@@ -25,43 +26,32 @@ public class BidPriceValidServiceImpl implements BidPriceValidService {
-                throw new InputMismatchException("BID_INPUT_MISMATCH", price);
+            int currentHighestBidPrice = currentHighestBid.getPrice();
+
+            if (newBidPrice > productHighestPrice || newBidPrice <= currentHighestBidPrice ||
```
newBidPrice > productHighestPrice ->
newBidPrice <= productHighestPrice 변경 부탁드립니다

gamsayeon replied:
구분을 위해 피드백 주신 newBidPrice으로 수정하도록 하겠습니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/config/MapperConfig.java:
```diff
@@ -11,7 +11,7 @@ public class MapperConfig {
     public ModelMapper modelMapper() {
         ModelMapper modelMapper = new ModelMapper();
-        modelMapper.getConfiguration().setMatchingStrategy(MatchingStrategies.LOOSE);
+        modelMapper.getConfiguration().setMatchingStrategy(MatchingStrategies.STRICT);
```
변경 사유를 주석으로 추가 부탁드립니다.

gamsayeon replied:
LOOSE은 필드 이름이 약간 다르더라도 매핑이 성공적으로 이루어집니다
 ex) DTO 에 user 와 Modle의 userId가 있을 경우 해당 전략을 사용하면 필드 이름이 정확히 일치 하지 않아도 매핑을 합니다.

STRICT은 필드 이름이 정확히 일치해야만 매핑이 성공적으로 이루어집니다.
ex) 위와 같은 예시로 DTO 에 user 와 Modle의 userId가 있을 경우 필드 이름이 일치하지 않으면 매핑이 수행되지 않기 때문입니다.

이러한 차이로 인해 해당 이름이 비슷한 필드 이름이 있습니다. 대표적으로 UserDTO와 User에서 UserDTO에 userId와 User의 기본키인 id와 이름이 비슷하여 필드 타입이 다르더라도 매핑을 시도하기에 해당 문제를 해결하고자 STRICT을 사용하여 정확히 일치한 필드로 매핑하도록 수정하였습니다.

ccommit-dev replied:
약간 다르더라도 매핑을 성공적으로 처리한다 -> 대소문자를가리지 않는다(?) 정도로 이해하면 될까요?

gamsayeon committed:
ENH: DLQ 설정 및 메시지 예외 처리 추가, 피드백 적용
WIP: Channel close Exception으로 인해 메시지 중복 상태 해결 중, 분산락인 Redission 락 관리(락 획득과 해제의 스레드가 달라 생기는 문제) 해결 중

gamsayeon committed:
ENH: todo 문제 우선 순위 큐 priority 설정 범위 해결

gamsayeon committed:
ENH: 락문제 해결(redis 저장소를 사용하여 락을 관리)

gamsayeon committed:
ENH: 락 갱신 코드 추가 및 락 문제 해결(비동기로 동작되어 메시지가 처리중에 있으나 DLQ의 Listener가 먼저 동작되는 문제가 있었으나 해결)

ccommit-dev commented on src/main/java/com/ccommit/auction_server/config/RabbitMQConfig.java:
```diff
@@ -108,6 +108,7 @@ public ConnectionFactory connectionFactory() {
         connectionFactory.setConnectionCacheSize(30);
         connectionFactory.setConnectionLimit(30);
+        connectionFactory.setConnectionTimeout(300000); // 연결 시간 초과 (밀리초)
         connectionFactory.setChannelCheckoutTimeout(3000000); // 채널 체크아웃 시간 (밀리초)
```
@gamsayeon 위의 두개의 차이를 정확하게 파악해서 알려주세요.

ccommit-dev replied:
CONNECTION TIME 시간도 profile config 설정파일로 빼두면 추후 live 배포시에 빌드배포가 아닌 파일만 변경배포 가능합니다. 참고부탁드립니다. 
[(Auto Configuration)](https://zion830.tistory.com/112)

gamsayeon committed:
ENH: 락 갱신 코드 추가 및 락 문제 해결(비동기로 동작되어 메시지가 처리중에 있으나 DLQ의 Listener가 먼저 동작되는 문제가 있었으나 해결)

ccommit-dev commented on src/main/java/com/ccommit/auction_server/service/serviceImpl/RabbitMQServiceImpl.java:
```diff
@@ -79,6 +80,9 @@ public class RabbitMQServiceImpl implements MQService {
     private static final int MAX_RETRY_COUNT = 10;
 
+    // Listener의 상태를 추적하는 플래그
+    private AtomicBoolean isListenerProcessing = new AtomicBoolean(false);
```
AtomicBoolean 변수를 사용한 이유를 댓글로 작성 부탁드립니다

gamsayeon replied:
해당 변수를 AtomicBoolean으로 설정한 이유는 멀티스레드 환경에서 메시지의 순서를 보장하기 위함입니다. 

구체적으로, 다음과 같은 시나리오를 처리하기 위해 사용됩니다.

1. 메시지를 처리하는 도중 치명적인 예외가 발생하면, 해당 메시지를 원본 큐(auction.queue)에서 Dead Letter Queue (DLQ)로 전송합니다.

2. DLQ로 메시지를 전송한 후, 원본 큐에서 같은 상품에 대한 입찰 정보가 남아 있을 경우, 해당 메시지의 처리가 DLQ에서 먼저 완료될 수 있습니다.
이 경우, 원본 큐에서 메시지가 처리 중일 때, 비동기로 동작하는 RabbitMQListener가 DLQ에서 메시지를 처리 완료할 수 있습니다. 

결과로는 원본 큐의 메시지가 여전히 처리 중일 때, DLQ의 메시지가 먼저 처리되고 락이 해제되므로 원본 큐의 메시지 처리 순서가 보장되지 않을 수 있습니다.

이를 해결하기 위해 AtomicBoolean을 사용하여 특정 상품에 대한 입찰 메시지에 대해 락을 걸어, 원본 큐에서 메시지를 처리하고 있을 때 DLQ에서 메시지를 처리하지 않도록 제어하였습니다. 이는 원본큐가 우선순위큐로 DLQ의 메시지가 처리가 완료되면 우선순위가 높은 메시지가 먼저 처리됨으로써 메시지의 순서를 보장할수 있도록 구성하였습니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/service/serviceImpl/RabbitMQServiceImpl.java:
```diff
@@ -153,16 +160,17 @@ public void dequeueMassage(Message message, Channel channel, @Header(AmqpHeaders
             logger.error("Handled Critical exception: " + ex.getMessage());
+        } finally {
+            // Listener가 메시지 처리를 완료하면 플래그를 false로 설정
+            isListenerProcessing.set(false);
```
finally 구문에서 해당 변수를 설정하지 않으면 안되는 사유를 주석에 상세하게 추가해주세요

gamsayeon replied:
DLQ Dequeue 메서드에 해당 변수를 체크 하여 DLQ의 리스너를 동작하도록 구성하였습니다.  

이는 설정하지 않을시 DLQ가 동작되지 않으며, DLQ의 메시지가 예외가 발생함에 따라 큐에 메세지가 쌓이게 됩니다.

finalliy에서 해당 변수를 설정한 이유로는 원본큐의 Listner을 동작하여 해당 메시지가 완전히 처리 됨을 나타내고자 해당 Finally에 동작하였습니다. 예외가 발생하거나 해당 메시지를 정상적으로 처리됨으로써 해당 메시지를 재처리 및 처리하고자

gamsayeon committed:
ENH: 코드 리펙토링

ccommit-dev commented on src/main/java/com/ccommit/auction_server/config/RabbitMQConfig.java:
```diff
@@ -112,4 +139,16 @@ public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
+        factory.setMessageConverter(jackson2JsonMessageConverter());
+        factory.setConcurrentConsumers(1);
+        factory.setMaxConcurrentConsumers(10);
+        factory.setPrefetchCount(10);  // 메시지 프리패치 수를 설정합니다.
```
프리패치 개념을 주석으로 추가 부탁드립니다

gamsayeon replied:
넵 알겠습니다!

ccommit-dev commented on src/main/java/com/ccommit/auction_server/config/RedissonConfig.java:
```diff
@@ -0,0 +1,25 @@
+    public RedissonClient redissonClient() {
+        Config config = new Config();
+        config.useSingleServer()
+                .setAddress("redis://" + redisHost + ":" + redisPort);
```
레디스를 singleserver와 클러스터로 설정했을떄의 장단점을 댓글로 정리해서 말씀 부탁드립니다

gamsayeon replied:
Single Server
장점
	- 설정이 간단: 단일 서버로 구성되기 때문에 설정과 유지 관리가 쉽습니다.
	- 비용 절감: 하나의 인스턴스만 사용하기 때문에 운영 비용이 적습니다.
	- 성능 우수: 모든 데이터와 작업이 한 곳에서 이루어지므로 네트워크 오버헤드가 적습니다.
단점
	- 단일 장애 지점(Single Point of Failure): 단일 서버로 구성되기에 서버에 문제가 생기면 서비스 전체가 중단될 수 있습니다.
	- 확장성 한계: 수평 확장이 어렵고, 트래픽 증가 시 성능 저하가 발생할 수 있습니다.
	- 백업 및 복구 어려움: 별도의 백업을 설정하지 않으면 장애 발생 시 데이터 손실 위험이 높습니다.

클러스터(Cluster)
장점
	- 고가용성: 데이터가 여러 노드에 분산되어 있어 한 노드가 다운되더라도 서비스 가용성이 유지됩니다.
	- 수평 확장 가능: 노드를 추가하여 저장 용량을 확장할 수 있습니다. 노드를 추가하여 부하를 분산할 수 있어 대규모 트래픽과 데이터를 처리할 수 있습니다.
	- 부하 분산: 데이터를 여러 노드에 분산 저장하며, 트래픽이 여러 노드로 분산 처리되어 부하를 분산시킬 수 있습니다.
단점
	- 설정 및 관리의 복잡성: 클러스터 구성은 각 노드 간의 통신, 데이터 분산(샤딩), 복제 설정 등이 필요합니다. 이러한 설정과 관리가 복잡하고 더 많은 관리 도구가 필요할 수 있습니다.
	- 비용 증가: 여러 서버를 운영해야 하므로 운영과 유지 관리 비용도 늘어납니다.
	- 데이터 일관성 문제: 데이터가 여러 노드에 걸쳐 분산 저장되기 때문에 데이터 동기화 과정에서 일관성 문제가 발생할 수 있습니다. 특히 네트워크 지연이나 노드 장애 시 데이터가 일시적으로 불일치할 수 있습니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/controller/ProductController.java:
```diff
@@ -144,13 +143,13 @@ public ResponseEntity<CommonResponse<String>> deleteProduct(@Parameter(hidden =
 
     // TODO: 성능테스트시 불필요한 Schedule 이기에 주석 처리 하였습니다.
-        @Scheduled(cron = "${auction.productState.scheduling.updateTime}")
+//    @Scheduled(cron = "${auction.productState.scheduling.updateTime}")
```
active profile 설정을 메서드 단위별로 되는지 조사해서 이경우에 적용해주시면 좋을듯 합니다

gamsayeon replied:
아 넵 알겠습니다.

해당 profile 어노테이션을 이용하여 메서드 단위별로 할수 있음을 확인하였습니다.

해당 구문은 스케줄링을 사용함에 있어 해당 구문을 pofile 별로 실행되어 사용됨을 확인했습니다.

ccommit-dev commented on src/main/java/com/ccommit/auction_server/validation/BidPriceValidator.java:
```diff
@@ -0,0 +1,55 @@
+    private static final Logger logger = LogManager.getLogger(BidPriceValidator.class);
+
+    public void validBidPrice(Long productId, int newBidPrice) {
+        Product product = productRepository.findByProductId(productId);
```
전반적으로 db를 table단위로 네트워크를 한번씩 갔다오는것 같은데 join 문으로 원하는 내용들을 한번에 조회하여 call횟수를 줄일 수 없을지 검토 부탁드립니다

gamsayeon replied:
아 넵 구조를 확인하여 join문으로 해결하도록 하겠습니다.

ccommit-dev reviewed (commented):
@gamsayeon 리뷰내용 확인 부탁드립니다

gamsayeon committed:
ENH: EnqueueController 로 분리하여 순환참조 해결(bidService <-> RabbitMQService)

gamsayeon committed:
ENH: bid validation bidService로 수정 및 join 문으로 리펙토링 (join문으로 수정시 각 model 별 외래키 설정및 join column 어노테이션 설정)
